drivers/platform/xilinx/timer.c

```diff
@@ -0,0 +1,385 @@
+{
+	struct xil_timer_desc *xil_timer_dev;
+
+	if(!desc)
+		return FAILURE;
+
+	xil_timer_dev = desc->extra;
+
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		XTmrCtr_Stop(xil_timer_dev->instance, xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		XScuTimer_Stop(xil_timer_dev->instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	free(xil_timer_dev->instance);
+	free(xil_timer_dev);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Start a timer.
+ * @param [in] desc - Pointer to the device handler.
+ * @return 0 in case of success, or negative error code in the following cases:
+ *         - the timer type is wrong.
+ */
+int32_t timer_start(struct timer_desc *desc)
+{
+	struct xil_timer_desc *xil_timer_dev = desc->extra;
+
+	xil_timer_dev->timer_counter_intr = 0;
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		XTmrCtr_Start(xil_timer_dev->instance, xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		XScuTimer_Start(xil_timer_dev->instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Stop a timer from counting.
+ * @param [in] desc - Pointer to the device handler.
+ * @return 0 in case of success, or negative error code in the following cases:
+ *         - the timer type is wrong.
+ */
+int32_t timer_stop(struct timer_desc *desc)
+{
+	struct xil_timer_desc *xil_timer_dev = desc->extra;
+
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		XTmrCtr_Stop(xil_timer_dev->instance, xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		XScuTimer_Stop(xil_timer_dev->instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the value of the counter register for the timer.
+ * @param [in]  desc    - Pointer to the device handler.
+ * @param [out] counter - Pointer to the counter value.
+ * @return 0 in case of success, or negative error code in the following cases:
+ *         - the timer type is wrong.
+ */
+int32_t timer_counter_get(struct timer_desc *desc, uint32_t *counter)
+{
+	struct xil_timer_desc *xil_timer_dev = desc->extra;
+
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		*counter = XTmrCtr_GetValue((XTmrCtr *)xil_timer_dev->instance,
+					    xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		*counter = XScuTimer_GetCounterValue((XScuTimer *)xil_timer_dev->
+						     instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the timer counter register value.
+ * @param [in] desc    - Pointer to the device handler.
+ * @param [in] new_val - The new value of the counter register.
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t timer_counter_set(struct timer_desc *desc, uint32_t new_val)
```

Comment:
This function is empty. 
Is it not enough to call the funtions `XTmrCtr_SetResetValue` and `XScuTimer_LoadTimer`?

Reply:
A good idea. I will implement it.